src/boot/efifirmware.c

```diff
@@ -0,0 +1,86 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efifirmware.h"
+#include "util.h"
+#include <endian.h>
+
+static bool efifw_validate_header(
+                const void *blob,
+                size_t blob_len,
+                const char **ret_fwid_off,
+                const char **ret_payload_off) {
+
+        if ((uintptr_t) blob % alignof(EfiFwHeader) != 0)
+                return false;
+
+        size_t base_sz = sizeof_field(EfiFwHeader, magic) +
+                sizeof_field(EfiFwHeader, header_len) +
+                sizeof_field(EfiFwHeader, fwid_len) +
+                sizeof_field(EfiFwHeader, payload_len);
```

Comment:
in this particular this calculation would be correct, but usually, due to potential padding it's safer to not sum up the size of fields, but rather take the offset of the next field.

specifically, maybe add a variable length array field `uint8_t payload[] _alignas_(uint64_t)` to the end of EfiFwHeader, whose job is to stand in for whatever payload comes next, in a byte addressable fashion, of unspecified size, but with maximum alignment (the latter is nice because it leaves all doors open for whatever comes next).

Then you could use `size_t base_sz = offsetof_field(EfiFwHeader, payload);` here.

Reply:
(just to underline this: codewise this would come to the exact same results, to some level i am just trying to be overly by-the-book with my suggestion)